JsonSchema/JsonSchema.cs

```diff
@@ -23,6 +24,7 @@ public class JsonSchema : IBaseDocument
 	
 	private readonly Dictionary<string, IJsonSchemaKeyword>? _keywords;
 	private readonly List<(DynamicScope Scope, SchemaConstraint Constraint)> _constraints = new();
+	private readonly ConcurrentDictionary<EvaluationOptions?, OptionsDerivedValues> _optionsDerivedValuesCache = new();
```

Comment:
Using the evaluation options as the key is basically a reference equality check (which may have hash collisions).  If a user is creating identical options objects, then the cache doesn't help anything.

Reply:
>Using the evaluation options as the key is basically a reference equality check

Yes. 

>(which may have hash collisions)

I don't follow. Can you elaborate?

>If a user is creating identical options objects, then the cache doesn't help anything.

Right. This is how STJ `Serialize`/`Deserialize` works with regards to `JsonSerializerOptions`, so there is precedent for that kind of behavior. Whether you consider it good design is another question I suppose. 

More generally, is this kind of caching something you'd be in favor of? I'd be happy to look into adding support for the case where the consumer passes different objects that are equal by value.

Reply:
I think my concern is more around memory usage when the user doesn't reuse options. I'd be holding onto those objects unnecessarily.

Reply:
I see. That's fair.

Maybe a dictionary is the wrong approach. My idea was to optimize for the case where the same options are re-used , and only a single cache entry is needed for that. I changed the code to only cache a single `OptionsDerivedValues` at a time. This satisfies my case while preventing the cache from growing over time. Does this seem like a better idea?

In order to get some idea of the size of the cache entry, I ran `GC.GetTotalMemory(true)` before and after creating and assigning it. The increase is 4952 bytes when using this schema. 

<details>
<summary>Schema</summary>

```
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/product.schema.json",
  "title": "Product",
  "description": "A product from Acme's catalog",
  "type": "object",
  "properties": {
    "productId": {
      "description": "The unique identifier for a product",
      "type": "integer"
    },
    "productName": {
      "description": "Name of the product",
      "type": "string"
    },
    "price": {
      "description": "The price of the product",
      "type": "number",
      "exclusiveMinimum": 0
    },
    "tags": {
      "description": "Tags for the product",
      "type": "array",
      "items": {
        "type": "string"
      },
      "minItems": 1,
      "uniqueItems": true
    },
    "dimensions": {
      "type": "object",
      "properties": {
        "length": {
          "type": "number"
        },
        "width": {
          "type": "number"
        },
        "height": {
          "type": "number"
        }
      },
      "required": [ "length", "width", "height" ]
    }
  },
  "required": [ "productId", "productName", "price" ]
}
```
</details>